compiler/rustc_trait_selection/src/traits/query/normalize.rs

```diff
@@ -64,6 +68,60 @@ impl<'cx, 'tcx> AtExt<'tcx> for At<'cx, 'tcx> {
             cache: SsoHashMap::new(),
             anon_depth: 0,
             universes: vec![],
+            mir_mode: false,
+        };
+
+        // This is actually a consequence by the way `normalize_erasing_regions` works currently.
+        // Because it needs to call the `normalize_generic_arg_after_erasing_regions`, it folds
+        // through tys and consts in a `TypeFoldable`. Importantly, it skips binders, leaving us
+        // with trying to normalize with escaping bound vars.
+        //
+        // Here, we just add the universes that we *would* have created had we passed through the binders.
+        //
+        // We *could* replace escaping bound vars eagerly here, but it doesn't seem really necessary.
+        // The rest of the code is already set up to be lazy about replacing bound vars,
+        // and only when we actually have to normalize.
+        if value.has_escaping_bound_vars() {
+            let mut max_visitor =
+                MaxEscapingBoundVarVisitor { outer_index: ty::INNERMOST, escaping: 0 };
+            value.visit_with(&mut max_visitor);
+            if max_visitor.escaping > 0 {
+                normalizer.universes.extend((0..max_visitor.escaping).map(|_| None));
+            }
+        }
+        let result = value.try_fold_with(&mut normalizer);
+        info!(
+            "normalize::<{}>: result={:?} with {} obligations",
+            std::any::type_name::<T>(),
+            result,
+            normalizer.obligations.len(),
+        );
+        debug!(
+            "normalize::<{}>: obligations={:?}",
+            std::any::type_name::<T>(),
+            normalizer.obligations,
+        );
+        result.map(|value| Normalized { value, obligations: normalizer.obligations })
+    }
+
+    /// Differs from `normalize` only in the way it evaluates mir constants.
+    fn normalize_for_mir<T>(&self, value: T) -> Result<Normalized<'tcx, T>, NoSolution>
```

Comment:
These two functions are thesame except for a Boolean. Make theBoolean an argument and only have one function. (Or have twowrapper functions inserting the bool as appropriate  )